await.go

```diff
@@ -0,0 +1,201 @@
+	// the cost at the point of adding a new entry.
+	//
+	// Once a waiter is added to this list, it belongs to the awaiter and the
+	// awaiter takes sole responsibility for closing the channel in the waiter.
+	waiters *list.List
+	// size and checkpoint keep track of the latest seen size and checkpoint as
+	// an optimization where the last seen value was already large enough.
+	size       uint64
+	checkpoint []byte
+}
+
+// Await blocks until the IndexFuture is resolved, and this new index has been
+// integrated into the log, i.e. the log has made a checkpoint available that
+// commits to this new index. When this happens, Await returns the index at
+// which the leaf has been added, and a checkpoint that commits to this index.
+//
+// This operation can be aborted early by cancelling the context. In this event,
+// or in the event that there is an error getting a valid checkpoint, an error
+// will be returned from this method.
+func (a *IntegrationAwaiter) Await(ctx context.Context, future IndexFuture) (uint64, []byte, error) {
+	i, err := future()
+	if err != nil {
+		return 0, nil, err
+	}
+	cp, err := a.await(ctx, i)
+	return i, cp, err
+}
+
+// pollLoop MUST be called in a goroutine when constructing an IntegrationAwaiter
+// and will run continually until its context is cancelled. It wakes up every
+// `pollPeriod` to check if there are clients blocking. If there are, it requests
+// the latest checkpoint from the log, parses the tree size, and releases all clients
+// that were blocked on an index smaller than this tree size.
+func (a *IntegrationAwaiter) pollLoop(ctx context.Context, readCheckpoint func(ctx context.Context) ([]byte, error), pollPeriod time.Duration) {
+	for {
+		select {
+		case <-ctx.Done():
+			klog.Info("IntegrationAwaiter exiting due to context completion")
+			return
+		case <-time.After(pollPeriod):
+			// It's worth this small lock contention to make sure that no unnecessary
+			// work happens in personalities that aren't performing writes.
+			a.mu.Lock()
+			hasClients := a.waiters.Front() != nil
+			a.mu.Unlock()
+			if !hasClients {
+				continue
+			}
+		}
+		// Note that for now, this releases all clients in the event of a single failure.
+		// If this causes problems, this could be changed to attempt retries.
+		rawCp, err := readCheckpoint(ctx)
+		if err != nil {
+			a.releaseClientsErr(fmt.Errorf("readCheckpoint: %v", err))
+			continue
+		}
+		// Parsing a checkpoint like this is only acceptable because we're in the same binary as the
+		// log implementation that generated it and thus we can safely assume it's a well formed and
+		// validly signed checkpoint. Anyone copying similar logic into client code will get hurt.
+		parts := bytes.SplitN(rawCp, []byte{'\n'}, 3)
+		if want, got := 3, len(parts); want != got {
+			a.releaseClientsErr(fmt.Errorf("invalid checkpoint: %q", rawCp))
+			continue
+		}
+		sizeStr := string(bytes.SplitN(rawCp, []byte{'\n'}, 4)[1])
```

Comment:
can this not use `parts[1]`?

Reply:
Good spot, I extracted the variable but didn't reuse it.